crates/revm/src/processor.rs

```diff
@@ -0,0 +1,509 @@
+use crate::{
+    database::StateProviderDatabase,
+    env::{fill_cfg_and_block_env, fill_tx_env},
+    eth_dao_fork::{DAO_HARDFORK_BENEFICIARY, DAO_HARDKFORK_ACCOUNTS},
+    into_reth_log,
+    stack::{InspectorStack, InspectorStackConfig},
+    state_change::post_block_balance_increments,
+};
+use reth_interfaces::{
+    executor::{BlockExecutionError, BlockValidationError},
+    Error,
+};
+use reth_primitives::{
+    Address, Block, BlockNumber, Bloom, ChainSpec, Hardfork, Header, PruneMode, PruneModes,
+    PrunePartError, Receipt, ReceiptWithBloom, TransactionSigned, H256, MINIMUM_PRUNING_DISTANCE,
+    U256,
+};
+use reth_provider::{
+    change::BundleStateWithReceipts, BlockExecutor, BlockExecutorStats, PrunableBlockExecutor,
+    StateProvider,
+};
+use revm::{
+    db::{states::bundle_state::BundleRetention, StateDBBox},
+    primitives::ResultAndState,
+    DatabaseCommit, State, EVM,
+};
+use std::{sync::Arc, time::Instant};
+use tracing::{debug, trace};
+
+/// Main block executor
+pub struct EVMProcessor<'a> {
+    /// The configured chain-spec
+    chain_spec: Arc<ChainSpec>,
+    evm: EVM<StateDBBox<'a, Error>>,
+    stack: InspectorStack,
+    /// The collection of receipts.
+    /// Outer vector stores receipts for each block sequentially.
+    /// The inner vector stores receipts that were not pruned.
+    receipts: Vec<Vec<Option<Receipt>>>,
+    /// First block will be initialized to ZERO
+    /// and be set to the block number of first block executed.
+    first_block: Option<BlockNumber>,
```

Comment:
this is in the sequence of blocks we execute? if so, prob fine:

```suggestion
    /// The block number of the first executed block.
    first_block: Option<BlockNumber>,
```

not sure what the significance of the first line is ("first block will be initialized to zero"), does this mean that it will be `Some(0)` at first any time you execute blocks? if so - why, and does this need to be an option in that case?

Reply:
Needs to be writen `None` not `ZERO`, the idea is when blocks start executing that we save first block, this is used for both running and later for creating the bundle state.